Implements automated release to SDKMAN

Karm opened the pull request:
```
$ ./mandrel-release.java sdkman -m /home/karm/workspaceRH/graal/graal -s Final -f Karm/graal
[jbang] Building jar...
[INFO] Current version is 21.3.1.1
[INFO] New version will be 21.3.1.2
[INFO] Releasing: mandrel-java11-linux-amd64-21.3.1.1-Final.tar.gz
[INFO] Released Response Code : 201
[INFO] Released Response Body : {"message":"Released: java 21.3.1.1.r11-mandrel for LINUX_64","status":201}
[INFO] Releasing: mandrel-java11-linux-aarch64-21.3.1.1-Final.tar.gz
[INFO] Released Response Code : 201
[INFO] Released Response Body : {"message":"Released: java 21.3.1.1.r11-mandrel for LINUX_ARM64","status":201}
[INFO] Releasing: mandrel-java11-windows-amd64-21.3.1.1-Final.zip
[INFO] Released Response Code : 201
[INFO] Released Response Body : {"message":"Released: java 21.3.1.1.r11-mandrel for WINDOWS_64","status":201}
[INFO] Releasing: mandrel-java17-linux-amd64-21.3.1.1-Final.tar.gz
[INFO] Released Response Code : 201
[INFO] Released Response Body : {"message":"Released: java 21.3.1.1.r17-mandrel for LINUX_64","status":201}
[INFO] Releasing: mandrel-java17-linux-aarch64-21.3.1.1-Final.tar.gz
[INFO] Released Response Code : 201
[INFO] Released Response Body : {"message":"Released: java 21.3.1.1.r17-mandrel for LINUX_ARM64","status":201}
[INFO] Releasing: mandrel-java17-windows-amd64-21.3.1.1-Final.zip
[INFO] Released Response Code : 201
[INFO] Released Response Body : {"message":"Released: java 21.3.1.1.r17-mandrel for WINDOWS_64","status":201}
[INFO] Promoting version 21.3.1.1.r11
[INFO] Promoted Response Code : 202
[INFO] Promoted Response Body : {"message":"Defaulted: java 21.3.1.1.r11-mandrel","status":202}
[INFO] Annoncing version 21.3.1.1.r11
[INFO] Annonce Response Code : 200
[INFO] Annonce Response Body : {"status":200,"id":"623342f06344590007b7da9c","message":"java 21.3.1.1.r11-mandrel available on SDKMAN! https://github.com/graalvm/mandrel/releases/"}
[INFO] Promoting version 21.3.1.1.r17
[INFO] Promoted Response Code : 202
[INFO] Promoted Response Body : {"message":"Defaulted: java 21.3.1.1.r17-mandrel","status":202}
[INFO] Annoncing version 21.3.1.1.r17
[INFO] Annonce Response Code : 200
[INFO] Annonce Response Body : {"status":200,"id":"623342f16344590007b7da9d","message":"java 21.3.1.1.r17-mandrel available on SDKMAN! https://github.com/graalvm/mandrel/releases/"}
```

Karm commented:
@zakkak  IMHO implemented. The release is visible on sdkman, *but*, I cannot verify it atm:
```
$ sdk install java 21.3.1.1.r17-mandrel

Downloading: java 21.3.1.1.r17-mandrel

In progress...

-#O=-#     #        #                                                                                                                                                                                             
Download has failed, aborting!

Can not install java 21.3.1.1.r17-mandrel at this time...

```

The reason it GitHub is mostly down: https://www.githubstatus.com/

So let's try to check it later....

Karm commented:
Hmm, not sure what is wrong.

```
$ sdk install java

Stop! java is not available. Possible causes:
 * is an invalid version
 * java binaries are incompatible with your platform
 * java has not been released yet

Tip: see all available versions for your platform:

  $ sdk list java
```

And we are also used in the docs as the default Java. Probably because we are the latest release...?
![image](https://user-images.githubusercontent.com/691097/158905524-c9d8f656-0bee-4e34-8abe-99e17f310f8b.png)

Older ones work just fine:
```
$ sdk install java 21.3.1.0.r17-mandrel

Downloading: java 21.3.1.0.r17-mandrel

In progress...

############################################################################################################################################################################################################ 100.0%############################################################################################################################################################################################################ 100.0%

Repackaging Java 21.3.1.0.r17-mandrel...

Done repackaging...

Installing: java 21.3.1.0.r17-mandrel
Done installing!


Setting java 21.3.1.0.r17-mandrel as default.
```

The one published by this automation doesn't:

```
$ sdk install java 21.3.1.1.r17-mandrel

Downloading: java 21.3.1.1.r17-mandrel

In progress...

-#O=-#     #        #                                                                                                                                                                                             
Download has failed, aborting!

Can not install java 21.3.1.1.r17-mandrel at this time...
```
I am not sure what could be wrong. It's seems to be really fool proof, e.g. my tested payload:

```
{
    "candidate": "java",
    "version": "21.3.1.1.r17",
    "vendor": "mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/download/mandrel-21.3.1.1-Final/mandrel-java17-linux-amd64-21.3.1.1-Final.tar.gz",
    "platform": "LINUX_64",
    "checksums": {
        "SHA-1": "adf1387196504dcfcc8a8080657f281d866d336d",
        "SHA-256": "768b608203cbde91d571ff03f9a9ce768a2bed6b62cffccaff2f19a0a7897edd"
    }
}
```

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+
+    /**
+     * One needs vendor credentials. Decipher those in:
+     * gpg --decrypt sdkman-vendor.json.asc > ~/.sdkman/vendor.json
```
It would possibly be even better if we could get the script to do the decrypt on the fly.

marc0der replied:
Will this be run on one of your local machines? The only reason I sent the credentials encrypted was that we don't like sending credentials in an email.

On the other hand, if this is an automated release script, please consider storing the key and token pair in GitHub secrets instead.

zakkak replied:
> Will this be run on one of your local machines? 

Yes, this will be run locally. So it's more about whether we would need to have the credentials stored in an unencrypted file (most probably on an encrypt disk though).

Karm replied:
It is a local workstation with an exclusive physical access and an encrypted drive. The file enjoys the same level of security, including filesystem permissions, as e.g. `~/.aws/credentials` or, for the matter of fact, `~/.ssh/id_rsa`.

zakkak commented:
Setting `sdkman_debug_mode` to `true` in `~./sdkman/etc/config` gives the following output:

```
$ sdk install java 21.3.1.1.r17-mandrel                                   
No update at this time. Using existing cache: activemq,ant,asciidoctorj,ballerina,bpipe,btrace,concurnas,connor,cuba,cxf,doctoolchain,flink,gaiden,gradle,gradleprofiler,grails,groovy,groovyserv,hadoop,http4k,infrastructor,java,jbake,jbang,jmc,jmeter,joern,jreleaser,karaf,kcctl,ki,kobweb,kotlin,kscript,layrry,leiningen,maven,mcs,micronaut,mulefd,mvnd,mybatis,pierrot,pomchecker,quarkus,sbt,scala,scalacli,schemacrawler,spark,springboot,sshoogr,taxi,test,tomcat,vertx,visualvm,webtau,znai
Not refreshing version cache now...
Validate java 21.3.1.1.r17-mandrel for linuxx64: valid
Validation URL: https://api.sdkman.io/2/candidates/validate/java/21.3.1.1.r17-mandrel/linuxx64
Get pre-installation hook: https://api.sdkman.io/2/hooks/pre/java/21.3.1.1.r17-mandrel/linuxx64
Copy remote pre-installation hook: /home/zakkak/.sdkman/tmp/hook_pre_java_21.3.1.1.r17-mandrel.sh
No Linux pre-install hook found for Java 21.3.1.1.r17-mandrel.
Completed pre-installation hook...

Downloading: java 21.3.1.1.r17-mandrel

In progress...

*   Trying 45.55.42.78:443...
* Connected to api.sdkman.io (45.55.42.78) port 443 (#0)
* ALPN, offering h2                                                                                                                                                                          
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/pki/tls/certs/ca-bundle.crt
*  CApath: none
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [25 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [5660 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [264 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: CN=*.sdkman.io
*  start date: May 30 00:00:00 2020 GMT
*  expire date: Apr 28 23:59:59 2022 GMT
*  subjectAltName: host "api.sdkman.io" matched cert's "*.sdkman.io"
*  issuer: C=GB; ST=Greater Manchester; L=Salford; O=Sectigo Limited; CN=Sectigo RSA Domain Validation Secure Server CA
*  SSL certificate verify ok.
} [5 bytes data]                                                                                                                                                                             
> GET /2/broker/download/java/21.3.1.1.r17-mandrel/linuxx64 HTTP/1.1
> Host: api.sdkman.io
> User-Agent: curl/7.79.1
> Accept: */*
> 
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [265 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):                                                                                                                                        
{ [265 bytes data]
* old SSL session ID is stale, removing
{ [5 bytes data]
* Mark bundle as not supporting multiuse
< HTTP/1.1 500 Internal Server Error
< Server: nginx/1.19.1
< Date: Fri, 18 Mar 2022 08:25:05 GMT
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host api.sdkman.io left intact

Downloaded binary to: /home/zakkak/.sdkman/tmp/java-21.3.1.1.r17-mandrel.bin (HTTP headers written to: /home/zakkak/.sdkman/var/metadata/java-21.3.1.1.r17-mandrel.headers)
Get post-installation hook: https://api.sdkman.io/2/hooks/post/java/21.3.1.1.r17-mandrel/linuxx64
Copy remote post-installation hook: /home/zakkak/.sdkman/tmp/hook_post_java_21.3.1.1.r17-mandrel.sh
A Linux post-install hook was found for Java 21.3.1.1.r17-mandrel.
Download has failed, aborting!

Can not install java 21.3.1.1.r17-mandrel at this time...
```

Notice the `< HTTP/1.1 500 Internal Server Error`.

My suspicion is that you "published" the new version while github was down and that somehow messed up the sdkman database update for the new version.

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+                        "        \"SHA-1\": \"" + sha1 + "\",\n" +
+                        "        \"SHA-256\": \"" + sha256 + "\"\n" +
+                        "    }\n" +
+                        "}";
```
Can you please add a debug print to dump the `releasePayload`?

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+                }
+                else
+                {
+                    error("File name " + assets[i].getName() + " does not match regexp " + filePattern.pattern() + " and that is unexpected. Skipping asset.");
```
```suggestion
                    warn("File name " + assets[i].getName() + " does not match regexp " + filePattern.pattern() + " and that is unexpected. Skipping asset.");
```

or

```suggestion
                    error("File name " + assets[i].getName() + " does not match regexp " + filePattern.pattern() + " and that is unexpected. Aborting.");
```
error is aborting the whole process it's not just skipping the asset.

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+        {
+            if (assets[i].exists() && assets[i + 1].exists() && assets[i + 2].exists())
+            {
+                final Matcher m = filePattern.matcher(assets[i].getName());
```
wouldn't it be better to make sure all 3 assets are matching the expected pattern?

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -579,6 +740,31 @@ private String getVersionFromRemoteManifest(String sourceURL, Pattern jdkVersion
+    private File[] assets(String fullVersion)
+    {
+        return new File[]{
+            // The order archive, sha1, sha256, matters.
```
Since we require a triplet, maybe we should just have the archive name here and append it with ".sha1" and ".sha256" when needed. This way we make it harder for copy paste mistakes to come up.

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+                    final String os = m.group("os").toUpperCase();
+                    final String arch = "amd64".equals(m.group("arch")) ? "64" : "ARM64";
+                    final String version = m.group("version") + ".r" + m.group("javaVersion").trim();
+                    versionsToPromote.add(version);
```
I don't think we want to do this blindly. E.g. we don't want to promote 21.3.1.1 over 22.0.0.2 even if the former is released after the latter. And we probably also want to control whether we will be promoting Java 11 or Java 17 variants.

zakkak replied:
Also why add the same version `n` times, where `n` is the number of supported OSs and Archs?

Karm replied:
> Also why add the same version `n` times, where `n` is the number of supported OSs and Archs?

That is supposed to address the situation when an asset is skipped, so it adds versions to a HashSet. There should be 2 of them (JDK 11 and JDK 17) at the end of the run.   We can probably simplify it. Unless we ditch the whole "promoting" altogether...

zakkak replied:
Yes, lets wait for Marco's reply!

marc0der replied:
Hi all, I've replied via email, but also doing it here for visibility. The Java candidate on SDKMAN is a shared space that all vendors use. The official default for Java is Temurin LTS (currently 17.0.2). Please do not set a default when publishing your release to SDKMAN, as this is something that we prefer to track and manage ourselves.

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+                warn("Skipping an incomplete asset " + assets[1].getName() + ". Make sure there are three files: archive, sha1 and sha256.");
+            }
+        }
+        for (String versionToPromote : versionsToPromote)
```
If passed more than one versions, it will "promote" all of them, essentially only promoting the last one. Am I missing something?

Karm replied:
I will wait for a hint from Marco.
My original understanding was that you have to promote all of them to transition from *candidate* to *released* or "default".

I might be mistaken though.

marc0der replied:
As before, please do not promote anything. We have a single default for _all_ Java versions, which is `17.0.2-tem` at this time.

Karm replied:
Done (code removed)

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+            info("Promoted Response Code : " + makeDefaultResponse.statusCode());
+            info("Promoted Response Body : " + makeDefaultResponse.body());
+
+            // Announce the released version
```
I haven't seen any other JDK variation being announced in https://twitter.com/sdkman_ so I am not sure we want this.

zakkak reviewed (changes requested):
Thanks for taking care of this @Karm !

Please find my suggestions/comments inline.

zakkak commented on release-scripts/mandrel-release.java:
```diff
@@ -479,6 +488,158 @@ private MandrelVersion getCurrentVersion()
+        final HttpClient hc = HttpClient.newHttpClient();
+        for (int i = 0; i + 3 <= assets.length; i = i + 3)
+        {
+            if (assets[i].exists() && assets[i + 1].exists() && assets[i + 2].exists())
```
Why do we need the files to exist locally since we are not uploading them at this step?

We just need the sha1 and sha256 hashes which we could fetch from github.

Karm replied:
I wanted to know they exist, so I put SDKMAN in this spot of the workflow.
Alternatively, we can check/download from GitHub Releases and if those URLs are valid, we run SDKMAN...
It largely depends on the expected workflow, if SDKMAN is something ran separately or if we run it in a one go.

Would you prefer fetching it from GitHub releases at this step instead of working with those locally stored files?

Karm replied:
Done

Karm commented:
@zakkak  debug & dryRun:
```
$ ./mandrel-release.java sdkman -m /home/karm/workspaceRH/graal/graal -s Final -f Karm/graal --verbose -D --sdkman-announce
[jbang] Building jar...
[INFO] Current version is 21.3.1.1
[INFO] New version will be 21.3.1.2
[INFO] Releasing: mandrel-java11-linux-amd64-21.3.1.1-Final.tar.gz
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r11",
    "vendor": "mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/download/mandrel-21.3.1.1-Final/mandrel-java11-linux-amd64-21.3.1.1-Final.tar.gz",
    "platform": "LINUX_64",
    "checksums": {
        "SHA-1": "ef1396f20e393d69924a60a83d0ca3b3e52d2211",
        "SHA-256": "4afccbc431c62f119c91a7bb077f885949f405e0a21be344edeb77dfeb0f5118"
    }
}
[INFO] Releasing: mandrel-java11-linux-aarch64-21.3.1.1-Final.tar.gz
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r11",
    "vendor": "mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/download/mandrel-21.3.1.1-Final/mandrel-java11-linux-aarch64-21.3.1.1-Final.tar.gz",
    "platform": "LINUX_ARM64",
    "checksums": {
        "SHA-1": "3eec38c1948e682af4a75ec7f7052076383f9b55",
        "SHA-256": "c5f75f0aa88dd60c7e0c8918f77f6b078dc2454545819d74ec085e0424cadf10"
    }
}
[INFO] Releasing: mandrel-java11-windows-amd64-21.3.1.1-Final.zip
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r11",
    "vendor": "mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/download/mandrel-21.3.1.1-Final/mandrel-java11-windows-amd64-21.3.1.1-Final.zip",
    "platform": "WINDOWS_64",
    "checksums": {
        "SHA-1": "a9c930ac92d875f72b72339bda54adcb33cefb2b",
        "SHA-256": "dae4910c9e804dfd44d158bc9e078a241f2f6f565de26bc32fc24fb5667b3f31"
    }
}
[INFO] Releasing: mandrel-java17-linux-amd64-21.3.1.1-Final.tar.gz
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r17",
    "vendor": "mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/download/mandrel-21.3.1.1-Final/mandrel-java17-linux-amd64-21.3.1.1-Final.tar.gz",
    "platform": "LINUX_64",
    "checksums": {
        "SHA-1": "adf1387196504dcfcc8a8080657f281d866d336d",
        "SHA-256": "768b608203cbde91d571ff03f9a9ce768a2bed6b62cffccaff2f19a0a7897edd"
    }
}
[INFO] Releasing: mandrel-java17-linux-aarch64-21.3.1.1-Final.tar.gz
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r17",
    "vendor": "mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/download/mandrel-21.3.1.1-Final/mandrel-java17-linux-aarch64-21.3.1.1-Final.tar.gz",
    "platform": "LINUX_ARM64",
    "checksums": {
        "SHA-1": "bfa341f17155940c586fb6e32c759ab07c96d7a1",
        "SHA-256": "83bc940a5bd1f8a9236c19dc4746be065a4c08cb095d8798fc715ea873d74e35"
    }
}
[INFO] Releasing: mandrel-java17-windows-amd64-21.3.1.1-Final.zip
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r17",
    "vendor": "mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/download/mandrel-21.3.1.1-Final/mandrel-java17-windows-amd64-21.3.1.1-Final.zip",
    "platform": "WINDOWS_64",
    "checksums": {
        "SHA-1": "1837b4f05d550a7e32330867d3b247198465ac35",
        "SHA-256": "45a0b454484af6bd4103ea033f781277093d2cd0d4b215ddd198ce71d9018218"
    }
}
[INFO] Promoting version 21.3.1.1.r11
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r11-mandrel",
    "vendor": "mandrel"
}
[INFO] Announcing version 21.3.1.1.r11
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r11-mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/"
}
[INFO] Promoting version 21.3.1.1.r17
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r17-mandrel",
    "vendor": "mandrel"
}
[INFO] Announcing version 21.3.1.1.r17
[DEBUG] {
    "candidate": "java",
    "version": "21.3.1.1.r17-mandrel",
    "url": "https://github.com/graalvm/mandrel/releases/"
}

```

Karm commented:
TL;DR from our conversation with Marco:
- remove announce, it won't be used
- remove promoting candidates; the semantic of the expression is different from what I though it to be
- add `visible=false` to hide older, surpassed versions of our Mandrel JDK, as in [this doc](https://github.com/sdkman/vendor-release/blob/master/src/test/resources/features/release_version_update.feature#L41)

aalmiray commented:
FWIW JReleaser can perform most of thew heavy lifting you're doing with the `mandrel-release.java` script

https://jreleaser.org

Karm commented:
Hello @marc0der,

Taking your recommendation about `"visible": false` to heart, I did what seems to be aligned with [this doc](https://github.com/sdkman/vendor-release/blob/master/src/test/resources/features/release_version_update.feature#L41).
There seems to be something odd though. Is the vendor string expected as a separate filed (contrary to the aforementioned `-zulu` example...)?   Thx for a hint.

```
$ ./mandrel-release.java sdkman --sdkman-version-invisible=21.3.1.0.r11  --verbose
[jbang] Building jar...
[DEBUG] {
  "candidate": "java",
  "version": "21.3.1.0.r11-mandrel",
  "platform": "LINUX_64",
  "visible": false
}
[INFO] Response Code : 400
[INFO] Response Body : {"message":"Version length exceeds 17 chars: 21.3.1.0.r11-mandrel","status":400}
[DEBUG] {
  "candidate": "java",
  "version": "21.3.1.0.r11-mandrel",
  "platform": "LINUX_ARM64",
  "visible": false
}
[INFO] Response Code : 400
[INFO] Response Body : {"message":"Version length exceeds 17 chars: 21.3.1.0.r11-mandrel","status":400}
[DEBUG] {
  "candidate": "java",
  "version": "21.3.1.0.r11-mandrel",
  "platform": "WINDOWS_64",
  "visible": false
}
[INFO] Response Code : 400
[INFO] Response Body : {"message":"Version length exceeds 17 chars: 21.3.1.0.r11-mandrel","status":400}
```


@zakkak 
I removed 
 * announce part of the script
 * "promoting" part of the script
 * changed the script to use GitHub releases instead of a local directory when processing artifacts

Karm commented:
@marc0der ...or we can drop the topic and wait for DISCO :smiley:

marc0der commented:
Well, DISCO integration is already fully operational (we performed the JDK 18 release with this new mechanism yesterday). I'd suggest just rolling with that instead. @eddumendelez can pick it up if you like.

Karm commented:
Merging as per our conversation on [SDKMAN Slack](https://sdkman.slack.com/archives/C02PF215HJ9/p1651076818463789?thread_ts=1647724447.439749&cid=C02PF215HJ9).
